Add liveview folder locations to Projectionist

zolrath opened the pull request:
Hello!

This adds the `live` folder to projectionist, though with Phoenix 1.7 lurking around the corner some other additions are likely useful with the modified directory structure.

mhanberg commented on lua/elixir/projectionist/init.lua:
```diff
@@ -32,14 +32,32 @@ local config = {
     },
+    ["lib/**/live/*_live.ex"] = {
+      type = "liveview",
+      alternate = "test/{dirname}/live/{basename}_live.exs",
```
```suggestion
      alternate = "test/{dirname}/live/{basename}_live_test.exs",
```

mhanberg commented on lua/elixir/projectionist/init.lua:
```diff
@@ -32,14 +32,32 @@ local config = {
+        "end",
+      },
+    },
+    ["test/**/live/*_live.exs"] = {
```
```suggestion
    ["test/**/live/*_live_test.exs"] = {
```

mhanberg commented on lua/elixir/projectionist/init.lua:
```diff
@@ -32,14 +32,32 @@ local config = {
+      alternate = "lib/{dirname}/live/{basename}_live.ex",
+      template = {
+        "defmodule {dirname|camelcase|capitalize}.{basename|camelcase|capitalize}LiveTest do",
+        "  use {dirname|camelcase|capitalize}.ConnCase, async: true",
```
Do LiveView tests use ConnCase?

zolrath replied:
Yep! Technically it looks like a generated test using the newest 1.7rc2 has a consistent:

```elixir
defmodule TestWeb.DocumentLiveTest do
  use TestWeb.ConnCase

  import Phoenix.LiveViewTest
```
it has ConnCase, though it doesn't set async to true and it has an additional LiveViewTest import as well.

mhanberg reviewed (changes requested):
I have a general question, I think the general naming convention for live views is something like `lib/my_app_web/live/foo_live/show.ex` and `MyAppWeb.FooLive.Show`.

If that is true, are you able to make the projection match that pattern? That was the problem  I had myself when i tried to add this in my own dotfiles.

zolrath commented:
Rebased and added the HTML/Component projections shown by German Velasco here: https://twitter.com/germsvel/status/1610283195018141696
Also changed LiveView tests to not add `async: true` as the default generators don't.

zolrath commented:
> I have a general question, I think the general naming convention for live views is something like `lib/my_app_web/live/foo_live/show.ex` and `MyAppWeb.FooLive.Show`.
> 
> If that is true, are you able to make the projection match that pattern? That was the problem I had myself when i tried to add this in my own dotfiles.

Most projects I've seen (LiveBeats as an example) tend to put their LiveViews in the root `live/` directory following the naming convention in this project.

https://github.com/fly-apps/live_beats/tree/master/lib/live_beats_web/live

It would be nice to support the folder containing the views components as seen in that repo, but I don't see a clear way to do that in projectionist at the moment.

mhanberg commented:
> Most projects I've seen (LiveBeats as an example) tend to put their LiveViews in the root live/ directory following the naming convention in this project

Interesting, the `phx.gen.live` mix task generates files that follow the `lib/phoenix_web/live/user_live/show.ex` pattern https://github.com/phoenixframework/phoenix/blob/a310102eb0e20b918624bb2323a6afb124fcaddd/test/mix/tasks/phx.gen.live_test.exs#L387

Probably alright tho, if people seem to disagree i assume they'll open an issue.

mhanberg commented on lua/elixir/projectionist/init.lua:
```diff
@@ -32,11 +32,73 @@ local config = {
+        "end",
+      },
+    },
+    ["lib/**/components/*_component.ex"] = {
```
this i would for sure just name `lib/**/components/*.ex`.

I don't think anyone would be naming their components in a way that would be `<TableComponent.render {@foo}>...</TableComponent.render>`

(well, at least I wouldn't ever name a component like that).

zolrath replied:
Hm, okay so you'd also change the module definition to 
```
        "defmodule {dirname|camelcase|capitalize}.{basename|camelcase|capitalize} do",
        "  use Phoenix.Component",
        "end",
```
or would you scope it differently?

Looking through other open source repos

https://github.com/BeaconCMS/beacon/blob/main/lib/beacon_web/live/admin/media_library_live/upload_form_component.ex
Dockyard seems to use _component on BeaconCMS

https://github.com/adoptoposs/adoptoposs/blob/develop/lib/adoptoposs_web/live/project_component.ex
This one seems to too!

https://github.com/qhwa/bonfire/blob/master/lib/bonfire_web/live/reading_state_componnet.ex
Same here

https://github.com/fremantle-industries/slurpee/blob/main/lib/slurpee_web/components/block_number_component.ex
Also here!

mhanberg replied:
We use Surface at my job, which actually has module components, and we don't name them with a "component" suffix 🤷.

But, I'll note that excluding the last link (which was last modified two years ago), those are all examples of LiveComponents, not function components.

I can add a PR that allows to easily override these defaults if that is what you'd like to see.

zolrath replied:
That is true, I suppose what might make sense would be another set of projections that detect `_component.ex` in the live directory and make live component tests?

mhanberg replied:
Yeah I did realize that there wasn't one for live components.

nor the new _json convention.

a PR for either/both of those would great!

mhanberg commented:
@zolrath thanks for getting this updated!

I'm going to pull it down and make sure it all works (on my machine), but other than my one comment it looks 💯 .

mhanberg commented:
> Rebased and added the HTML/Component projections shown by German Velasco here: https://twitter.com/germsvel/status/1610283195018141696 Also changed LiveView tests to not add `async: true` as the default generators don't.

Oh I didn't see this comment before I left mine about the Component projection. I think I still disagree, even with German's testimony.

mhanberg commented:
I pulled it down and fixed  several bugs, but it's good to go now.

Thank you so much!
